### DISCUSSION TITLE
Support Dashboards as code and user created dashboards in tandem

### DISCUSSION THREAD

**Discussion by jgbernalp** [perses/perses#1182] [Design Discussions]
Context: 
Dashboards could be created as code via CRDs, but users might also want to create their own dashboards which can be stored using a different storage strategy such as a database. 
Proposed solution:

- Use 2 instances of perses, one that can read CRDs and the other can use the database for read-write user created dashboards. The risks is that clients would need to know upfront which perses to contact based on the type of dashboard. This will also conflict with dashboard uniqueness as 2 perses can create a dashboard with the same id.

**Comment by Nexucis** [comment]
> This will also conflict with dashboard uniqueness as 2 perses can create a dashboard with the same id.

Actually it can't happen with two different instance.
It will happen if you have a single instance reading CR and having a database to create other dashboards.

If you have an instance that is using CRDs and at the same time it's connected to a database, that means the project object will be mapped to the k8s namespace resources. So if at some point for example, you removed a namespace called `foo` and then you create a project `foo` in Perses and then the namespace is back in k8s. You would have two projects with the same name and it will be impossible to know which one to use.  The unicity will be broken.

**Reply by jgbernalp** [reply]
In that sense then we will have 2 sources of truth when handling multiple data storages in a single Perses. would it be better to have then a database and sync the data from CRDs or files into the database as @nicolastakashi suggested? Maybe a command in percli to import data?

**Comment by nicolastakashi** [comment]
A third option would be we allow save dashboards when running on Kubernetes through the UI and leverage admission webhooks to lint and save the dashboard on external sources, with this we can achieve both of the best worlds in my view.

Cause when we post the dashboard JSON through the UI to Kubernetes API the Admission Webhook will be triggered and we can apply the lint we want and store this on other data sources such as Github.

To not mention we open an integration door where the use can build their own admission webhook to handle when a Dashboard CRD is created.

**Reply by jgbernalp** [reply]
Related issue: https://github.com/perses/perses-operator/issues/119

**Comment by Nexucis** [comment]
I believe with the perses-operator, this should be possible now. CRs are copied to the Perses database and then use can create others directly through the UI.

**Comment by Nexucis** [comment]
closing this discussion as it is solved thanks to the `perses-operator`
